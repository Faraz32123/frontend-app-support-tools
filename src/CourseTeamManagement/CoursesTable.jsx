import React, { useEffect, useState, useRef } from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';
import {
  Button,
  Form,
  DataTable,
  Badge,
  Dropdown,
} from '@openedx/paragon';
import { ArrowDropDown } from '@openedx/paragon/icons';
import PropTypes from 'prop-types';
import messages from './messages';
import SortableHeader from './customSortableHeader';
import TableActions from './customTableActions';

export default function CoursesTable({ username, userCourses }) {
  const intl = useIntl();
  let userCoursesData = userCourses;

  // Change role null to 'null' for sorting to working correctly
  // As we are considering 'null' to appear as staff with disabled dropdown
  userCoursesData = userCoursesData.map((course) => ({
    ...course,
    role: course.role === null ? 'null' : course.role,
  }));
  const [search, setSearch] = useState('');
  const [status, setStatus] = useState('');
  const [org, setOrg] = useState('');
  const [sortBy, setSortBy] = useState([]);
  // Track roles for each course row, keyed by run
  const [rowRoles, setRowRoles] = useState(() => userCoursesData.reduce((acc, row) => {
    acc[row.run] = row.role;
    return acc;
  }, {}));
  const searchInputRef = useRef(null);

  const sortedAndFilteredData = React.useMemo(() => {
    let data = userCoursesData.map((row) => ({
      ...row,
      role: rowRoles[row.run] !== undefined ? rowRoles[row.run] : 'null',
    }));

    // Manual Filtering for all columns in single search box
    data = data.filter(
      (row) => (row.course_name.toLowerCase().includes(search.toLowerCase())
          || row.number.toLowerCase().includes(search.toLowerCase())
          || row.run.toLowerCase().includes(search.toLowerCase())
      )
      && (status === '' || row.status === status)
      && (org === '' || row.org === org),
    );

    // Manual sorting
    if (sortBy.length > 0) {
      const { id, desc } = sortBy[0];
      data = [...data].sort((a, b) => {
        const aValue = a[id];
        const bValue = b[id];
        if (aValue < bValue) { return desc ? 1 : -1; }
        if (aValue > bValue) { return desc ? -1 : 1; }
        return 0;
      });
    }

    return data;
  }, [search, status, org, userCoursesData, rowRoles, sortBy]);

  // Focus the search input when the component mounts
  useEffect(() => {
    if (searchInputRef.current) {
      searchInputRef.current.focus();
    }
  }, []);

  // Custom checkbox selection state, keyed by run
  const [checkedRows, setCheckedRows] = useState(() => {
    const initial = {};
    userCoursesData.forEach((row) => {
      if (row.role === 'staff' || row.role === 'instructor') {
        initial[row.run] = true;
      }
    });
    return initial;
  });

  const handleCheckboxChange = (runId) => {
    setCheckedRows((prev) => ({ ...prev, [runId]: !prev[runId] }));
  };

  // Select all/clear all logic for header checkbox
  const headerCheckboxRef = useRef(null);
  const allRowIds = sortedAndFilteredData.map((row) => row.run);
  const numChecked = allRowIds.filter((id) => checkedRows[id]).length;
  const allChecked = numChecked === allRowIds.length && allRowIds.length > 0;
  const someChecked = numChecked > 0 && numChecked < allRowIds.length;
  useEffect(() => {
    if (headerCheckboxRef.current) {
      headerCheckboxRef.current.indeterminate = someChecked && !allChecked;
    }
  }, [someChecked, allChecked, numChecked, sortBy]);

  const handleHeaderCheckboxChange = () => {
    if (allChecked) {
      // Clear all
      setCheckedRows((prev) => {
        const updated = { ...prev };
        allRowIds.forEach((id) => {
          updated[id] = false;
        });
        return updated;
      });
    } else {
      // Select all
      setCheckedRows((prev) => {
        const updated = { ...prev };
        allRowIds.forEach((id) => {
          updated[id] = true;
        });
        return updated;
      });
    }
  };

  // Generate unique orgs for org filterChoices
  const orgFilterChoices = Array.from(new Set(userCoursesData.map((row) => row.org)))
    .filter(Boolean)
    .map((organization) => ({ name: organization, value: organization }));

  const formatStatus = ({ value }) => (
    <Badge
      className="course-team-management-course-status-badge"
      variant={value === 'active' ? 'success' : 'light'}
    >
      {value === 'active'
        ? intl.formatMessage(messages.activeCoursesFilterLabel)
        : intl.formatMessage(messages.archivedCoursesFilterLabel)}
    </Badge>
  );

  const formatCourseName = ({ row }) => {
    const courseLink = row.original.course_url;
    const courseName = row.original.course_name;
    return (
      <a href={courseLink} target="_blank" rel="noopener noreferrer">
        {courseName}
      </a>
    );
  };

  const formatRole = ({ row }) => {
    const runId = row.original.run;
    const value = rowRoles[runId];
    // If role is 'null', default to staff for display only
    const displayValue = value === 'null' ? 'staff' : value;
    let title = 'Staff';
    if (displayValue === 'instructor') { title = 'Admin'; }
    // Enable dropdown if checkbox is checked, otherwise disable
    const isChecked = !!checkedRows[runId];
    const isDisabled = !isChecked;
    return (
      <Dropdown>
        <Dropdown.Toggle
          as="button"
          className="course-team-management-role-col-dropdown"
          data-testid={`role-dropdown-${runId}`}
          variant="outline-primary"
          disabled={isDisabled}
          style={{
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'space-between',
            padding: '2px 8px',
            width: '95px',
            borderRadius: '6px',
          }}
        >
          <span>{title}</span>
          <ArrowDropDown className="ml-2" />
        </Dropdown.Toggle>

        <Dropdown.Menu placement="top">
          <Dropdown.Item
            data-testid={`role-dropdown-item-staff-${runId}`}
            eventKey="staff"
            active={displayValue === 'staff'}
            onClick={() => setRowRoles((prev) => ({ ...prev, [runId]: 'staff' }))}
          >
            {intl.formatMessage(messages.statusStaffFilterLabelChoice)}
          </Dropdown.Item>
          <Dropdown.Item
            data-testid={`role-dropdown-item-instructor-${runId}`}
            eventKey="instructor"
            active={displayValue === 'instructor'}
            onClick={() => setRowRoles((prev) => ({ ...prev, [runId]: 'instructor' }))}
          >
            {intl.formatMessage(messages.statusAdminFilterLabelChoice)}
          </Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
    );
  };

  // Custom filter UI for tableActions
  const tableActions = (
    <TableActions
      search={search}
      setSearch={setSearch}
      searchInputRef={searchInputRef}
      status={status}
      setStatus={setStatus}
      org={org}
      setOrg={setOrg}
      orgFilterChoices={orgFilterChoices}
      setRowRoles={setRowRoles}
      sortedAndFilteredData={sortedAndFilteredData}
      checkedRows={checkedRows}
      userCoursesData={userCoursesData}
    />
  );
  // eslint-disable-next-line react/no-unstable-nested-components
  const CheckboxHeader = () => (
    <Form.Check
      ref={headerCheckboxRef}
      type="checkbox"
      checked={allChecked}
      onChange={handleHeaderCheckboxChange}
      aria-label="Select all rows"
      style={{ marginLeft: 8 }}
    />
  );
  // eslint-disable-next-line react/no-unstable-nested-components
  const CheckboxCell = ({ row }) => (
    <Form.Check
      type="checkbox"
      checked={!!checkedRows[row.original.run]}
      onChange={() => handleCheckboxChange(row.original.run)}
      aria-label={`Select row for ${row.original.course_name}`}
      style={{ marginLeft: 8 }}
    />
  );
  const sortableHeader = (id, label) => (
    <SortableHeader
      id={id}
      label={label}
      sortBy={sortBy}
      setSortBy={setSortBy}
    />
  );

  return (
    <div className="course-team-management-courses-table">
      <header className="course-team-management-courses-table-title">
        {intl.formatMessage(messages.courseAccessForUsernameTitle, {
          username,
        })}
      </header>{' '}
      {}
      <div className="course-team-management-courses-table-description">
        <p>{intl.formatMessage(messages.courseAccessDescription)}</p>
      </div>
      <DataTable
        isSortable
        manualSortBy
        sortBy={sortBy}
        onSortByChange={setSortBy}
        isFilterable
        itemCount={userCoursesData.length}
        tableActions={[tableActions]}
        data={sortedAndFilteredData}
        columns={[
          {
            Header: CheckboxHeader,
            accessor: 'checkbox',
            disableFilters: true,
            disableSortBy: true,
            id: 'checkbox',
            Cell: CheckboxCell,
          },
          {
            Header: sortableHeader('course_name', messages.tableHeaderNameLabel),
            accessor: 'course_name',
            id: 'course_name',
            Cell: formatCourseName,
            disableFilters: true,
          },
          {
            Header: sortableHeader('number', messages.tableHeaderNumberLabel),
            accessor: 'number',
            id: 'number',
            disableFilters: true,
          },
          {
            Header: sortableHeader('run', messages.tableHeaderRunLabel),
            accessor: 'run',
            id: 'run',
            disableFilters: true,
          },
          {
            Header: sortableHeader('status', messages.tableHeaderAllCoursesStatusLabel),
            accessor: 'status',
            id: 'status',
            Cell: formatStatus,
            disableFilters: true,
          },
          {
            Header: sortableHeader('role', messages.tableHeaderRoleLabel),
            accessor: 'role',
            id: 'role',
            Cell: formatRole,
            disableFilters: true,
          },
        ]}
        getRowId={(row) => row.run}
      >
        <DataTable.TableControlBar />
        <div style={{ height: 450, overflow: 'auto' }}>
          {sortedAndFilteredData.length > 0 && <DataTable.Table />}
          {sortedAndFilteredData.length === 0 && (
            <div className="pgn__data-table-empty">{intl.formatMessage(messages.noResultsFoundForTable)}</div>
          )}
        </div>
        {sortedAndFilteredData.length > 0 && <DataTable.TableFooter />}
      </DataTable>
      <div className="py-4 my-2 d-flex justify-content-end align-items-center">
        <Button variant="brand" style={{ width: '8%' }}>
          {intl.formatMessage(messages.saveButtonLabel)}
        </Button>
      </div>
    </div>
  );
}

CoursesTable.propTypes = {
  username: PropTypes.string.isRequired,
  userCourses: PropTypes.string.isRequired,
  row: PropTypes.shape({
    run: PropTypes.string.isRequired,
    original: PropTypes.shape({
      run: PropTypes.string.isRequired,
      course_name: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};
