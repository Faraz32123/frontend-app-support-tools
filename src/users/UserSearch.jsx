import React, {
  useRef, useCallback,
} from 'react';
import PropTypes from 'prop-types';

import { Button, Form, Icon } from '@openedx/paragon';
import { Search } from '@openedx/paragon/icons';

export default function UserSearch({ userIdentifier, searchHandler, isOnCourseTeamPage }) {
  const searchRef = useRef();

  const submit = useCallback((event) => {
    searchHandler(searchRef.current.value);
    event.preventDefault();
    return false;
  });

  return (
    <section className={`mb-3${!isOnCourseTeamPage ? ' px-2' : ''}`}> 
      <Form className={isOnCourseTeamPage ? "course-team-management-user-search-form" : ""}>
        <Form.Row className={isOnCourseTeamPage ? "course-team-management-user-search-form-row" : ""}>
          {!isOnCourseTeamPage && <Form.Label htmlFor="userIdentifier" className="my-auto">Username, Email or LMS User ID</Form.Label>}
          <Form.Control
            ref={searchRef}
            className={`${!isOnCourseTeamPage ? "" : "flex-grow-1 "}ml-1 mr-1`}
            name="userIdentifier"
            defaultValue={userIdentifier}
            floatingLabel={isOnCourseTeamPage ? "Username or email" : ""}
            trailingElement={isOnCourseTeamPage ? <Icon src={Search} /> : ""}
          />
          <Button type="submit" onClick={submit} variant="primary">Search</Button>
        </Form.Row>
      </Form>
    </section>
  );
}

UserSearch.propTypes = {
  userIdentifier: PropTypes.string,
  searchHandler: PropTypes.func.isRequired,
  isOnCourseTeamPage: PropTypes.bool,
};

UserSearch.defaultProps = {
  userIdentifier: '',
  isOnCourseTeamPage: false,
};
