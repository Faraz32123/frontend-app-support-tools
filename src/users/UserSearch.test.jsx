import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import UserSearch from './UserSearch';
import { intlProviderWrapper } from '../CourseTeamManagement/CoursesTable.test';

describe('User Search Page', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = { userIdentifier: '', searchHandler: jest.fn() };
    wrapper = mount(intlProviderWrapper(<UserSearch {...props} />));
  });

  describe('renders correctly', () => {
    it('with correct user identifier', () => {
      expect(wrapper.find('input').prop('defaultValue')).toEqual(
        props.userIdentifier,
      );
    });
    it('with correct default user identifier', () => {
      delete props.userIdentifier;
      const userSearchwrapper = mount(intlProviderWrapper(<UserSearch {...props} />));

      expect(userSearchwrapper.find('input').prop('defaultValue')).toEqual('');
    });
    it('with submit button', () => {
      expect(wrapper.find('button')).toHaveLength(1);
      expect(wrapper.find('button').text()).toEqual('Search');
    });

    it('when submit button is clicked', () => {
      const searchProps = { userIdentifier: 'staff', searchHandler: jest.fn() };
      const userSearchwrapper = mount(intlProviderWrapper(<UserSearch {...searchProps} />));

      userSearchwrapper.find('button').simulate('click');

      expect(searchProps.searchHandler).toHaveBeenCalledWith(
        searchProps.userIdentifier,
      );
    });

    it('matches snapshot', () => {
      const tree = renderer.create(wrapper).toJSON();

      expect(tree).toMatchSnapshot();
    });
    it('matches snapshot for course team page', () => {
      const courseTeamWrapper = renderer
        .create(intlProviderWrapper(<UserSearch {...props} isOnCourseTeamPage />))
        .toJSON();
      expect(courseTeamWrapper).toMatchSnapshot();
    });
  });

  describe('course team page behaviors (unsaved changes modal)', () => {
    beforeEach(() => {
      sessionStorage.clear();
    });

    it('opens modal instead of searching when unsaved changes exist', () => {
      const searchHandler = jest.fn();
      const username = 'tester';
      sessionStorage.setItem(`${username}hasUnsavedChanges`, 'true');

      const ctmWrapper = mount(intlProviderWrapper(
        <UserSearch userIdentifier="foo" searchHandler={searchHandler} isOnCourseTeamPage username={username} />,
      ));

      ctmWrapper.find("input[name='userIdentifier']").instance().value = 'nextQuery';
      ctmWrapper.find('button').simulate('click');

      expect(searchHandler).not.toHaveBeenCalled();
      const modal = ctmWrapper.find('CustomLeaveModalPopup');
      expect(modal.prop('isOpen')).toBe(true);
    });

    it('confirming modal proceeds with search and clears unsaved changes flag', () => {
      const searchHandler = jest.fn();
      const username = 'tester2';
      sessionStorage.setItem(`${username}hasUnsavedChanges`, 'true');

      const ctmWrapper = mount(intlProviderWrapper(
        <UserSearch userIdentifier="foo" searchHandler={searchHandler} isOnCourseTeamPage username={username} />,
      ));

      ctmWrapper.find("input[name='userIdentifier']").instance().value = 'confirmedQuery';
      ctmWrapper.find('button').simulate('click');

      const modal = ctmWrapper.find('CustomLeaveModalPopup');
      modal.prop('onConfirm')();
      ctmWrapper.update();

      expect(searchHandler).toHaveBeenCalledWith('confirmedQuery');
      expect(sessionStorage.getItem(`${username}hasUnsavedChanges`)).toBe('false');
      expect(ctmWrapper.find('CustomLeaveModalPopup').prop('isOpen')).toBe(false);
    });

    it('canceling modal closes it without searching and keeps flag intact', () => {
      const searchHandler = jest.fn();
      const username = 'tester3';
      sessionStorage.setItem(`${username}hasUnsavedChanges`, 'true');

      const ctmWrapper = mount(intlProviderWrapper(
        <UserSearch userIdentifier="foo" searchHandler={searchHandler} isOnCourseTeamPage username={username} />,
      ));

      ctmWrapper.find('button').simulate('click');

      const modal = ctmWrapper.find('CustomLeaveModalPopup');
      modal.prop('onCancel')();
      ctmWrapper.update();

      expect(searchHandler).not.toHaveBeenCalled();
      expect(ctmWrapper.find('CustomLeaveModalPopup').prop('isOpen')).toBe(false);
      expect(sessionStorage.getItem(`${username}hasUnsavedChanges`)).toBe('true');
    });

    it('searches immediately when no unsaved changes exist', () => {
      const searchHandler = jest.fn();
      const username = 'tester4';
      sessionStorage.setItem(`${username}hasUnsavedChanges`, 'false');

      const ctmWrapper = mount(intlProviderWrapper(
        <UserSearch userIdentifier="foo" searchHandler={searchHandler} isOnCourseTeamPage username={username} />,
      ));

      ctmWrapper.find("input[name='userIdentifier']").instance().value = 'directQuery';
      ctmWrapper.find('button').simulate('click');

      expect(searchHandler).toHaveBeenCalledWith('directQuery');
      expect(ctmWrapper.find('CustomLeaveModalPopup').prop('isOpen')).toBe(false);
    });
  });
});
