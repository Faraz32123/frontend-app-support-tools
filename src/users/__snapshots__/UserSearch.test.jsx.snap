// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Search Page renders correctly matches snapshot 1`] = `
<section
  className="mb-3 px-2"
>
  <form
    className=""
  >
    <div
      className="form-row"
    >
      <label
        className="pgn__form-label my-auto"
        htmlFor="userIdentifier"
      >
        Username, Email or LMS User ID
      </label>
      <div
        className="pgn__form-control-decorator-group ml-1 mr-1"
      >
        <input
          className="form-control"
          defaultValue=""
          name="userIdentifier"
          onBlur={[Function]}
          onChange={[Function]}
        />
        
        
      </div>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</section>
`;

exports[`User Search Page renders correctly matches snapshot for course team page 1`] = `
<section
  className="mb-3"
>
  <form
    className="course-team-management-user-search-form"
  >
    <div
      className="course-team-management-user-search-form-row form-row"
    >
      <div
        className="pgn__form-control-decorator-group has-appended-node has-trailing-element has-floating-label flex-grow-1 ml-1 mr-1"
      >
        <input
          className="form-control"
          defaultValue=""
          name="userIdentifier"
          onBlur={[Function]}
          onChange={[Function]}
        />
        <div
          className="pgn__form-control-decorator pgn__form-control-decorator-trailing"
        >
          <span
            className="pgn__icon"
          >
            <svg
              aria-hidden={true}
              fill="none"
              focusable={false}
              height={24}
              role="img"
              viewBox="0 0 24 24"
              width={24}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0 0 16 9.5 6.5 6.5 0 1 0 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5Zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </div>
        <div
          className="pgn__form-control-floating-label"
        >
          <div
            className="pgn__form-control-floating-label-content"
          >
            <label
              className="pgn__form-control-floating-label-text"
            >
              Username or email
            </label>
          </div>
        </div>
      </div>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</section>
`;
